### PR TITLE
MACE Preoptimisation for FHI-aims

### PR DESCRIPTION
**Pull request by GabrielBram**
A workflow intended for pre-optimising structures and NEB transition state paths with the MACE family of forcefields. 

Functionality invoked through the standard aims.optimise and aims.ts_search commands, with a user supplied MACE pre-optimiser to the ReactAims object, eg.,

```
mace_params = {'model': 'large',
               'dispersion': False,
               'default_dtype': 'float64',
               'device': 'cpu'}

reactor = ReactAims(params, basis_set, hpc)
reactor.MaceReact_Preoptimiser = ReactMACE(mace_params, force_field="mace_mp")

atoms = reactor.aims_optimise(atoms, 0.02, relax_unit_cell=False, restart=True, mace_preopt=True)[0]

```

### CODE REVIEW

**Comment by ikowalec** [comment]
This currently breaks the restarts on task-farming

**Comment by GabrielBram** [comment]
Dw, was just seeing how it behaved in the test infrastructure. By the looks of it, it's not recognising the restart as converged and attempts to run an FHI-aims simulation even though it can't. Why it's doing that is another matter.

...It also breaks itself. 

Hmm. Will request the review when everything is in working order.

**Comment by GabrielBram** [comment]
Magical - it works on my machine but not on the remote branch. Something stupid must have happened.

**Comment by GabrielBram** [comment]
Hello! This is somewhat ready - I do not like some of the solutions herein (e.g., the approach of re-attaching calculators after the MACE run), and some of the restart logic needs re-working. However, just to get the branch out there and to field suggestions, I've submitted a request for review.

**Comment by GabrielBram** [comment]
(Please don't all code review - just adding enough people to get the merge through).

**Review by kush2803 — APPROVED** [review]

File: carmm/run/workflows/react.py [hunk 1]
```diff
@@ -548,6 +609,120 @@ def vibrate(self, atoms: Atoms, indices: list, read_only=False):
...
+
+        self._MaceReactor = MaceReact
+
+    def _mace_preoptimise(self, atoms: Atoms, fmax, relax_unit_cell, optimiser, opt_kwargs = {}):
+        """
+        Internal function for preoptimisation of structures with a pre-set MACE calculator
+
+        Args:
+            atoms: Atoms object
+            fmax: flaot
```
Comment (logsdail) [inline carmm/run/workflows/react.py:639]: Typo.

Comment (logsdail) [reply carmm/run/workflows/react.py:639]: Consider making it clear what these are for the next user - or point to where clearly defined.

File: carmm/run/workflows/react.py [hunk 2]
```diff
@@ -548,6 +609,120 @@ def vibrate(self, atoms: Atoms, indices: list, read_only=False):
...
+        preopt_atoms.calc = None
+
+        return preopt_atoms
+
+    def _mace_preoptimise_ts(self, initial, final, fmax, n, interpolation, input_check, max_steps=200):
+        """
+        Internal function for preoptimisation of NEB with a pre-set MACE calculator
+
+        Args:
+            initial: Atoms object
```
Comment (logsdail) [inline carmm/run/workflows/react.py:666]: As above, if same as somewhere else then say so?

**Review by logsdail — COMMENTED** [review]
Generally looks very reasonable _but_ I’m not a user, just looking at the code as written!

File: carmm/run/workflows/react.py (hunk 1 above)
Comment (GabrielBram) [reply carmm/run/workflows/react.py]: Cheers! Both resolved.

File: carmm/run/workflows/react.py (hunk 2 above)
Comment (GabrielBram) [reply carmm/run/workflows/react.py:670]: Added additional comments to explain, thank you!

**Pull request merged**